core/middlewares/with-routes.ts

```diff
@@ -223,6 +223,23 @@ const clearLocaleFromPath = (path: string, locale: string) => {
   return path;
 };
 
+function normalizeForCompare(url: URL): string {
+  const trailingSlashDisabled = process.env.TRAILING_SLASH === 'false';
+
+  if (trailingSlashDisabled && url.pathname !== '/' && url.pathname.endsWith('/')) {
+    return `${url.pathname.replace(/\/+$/, '')}${url.search}`;
+  }
+
+  if (!trailingSlashDisabled && !url.pathname.endsWith('/')) {
+    return `${url.pathname}/${url.search}`;
+  }
+
+  return `${url.pathname}${url.search}`;
+}
```

Comment:
I'm just posting my AI response here. I prefer the flow of code + handling the lowercasing of mine, but yours deals with the search params and utilizes a `URL` object. Might want to pick and choose some things between the two.

```ts
const trailingSlashEnabled = process.env.TRAILING_SLASH !== 'false';

const normalizeUrlForComparison = (path: string): string => {
  let normalized = path.toLowerCase();

  if (trailingSlashEnabled) {
    if (normalized !== '/' && !normalized.endsWith('/')) {
      normalized += '/';
    }

    return normalized;
  }

  if (normalized !== '/' && normalized.endsWith('/')) {
    normalized = normalized.slice(0, -1);
  }

  return normalized;
};
```

Reply:
I'm not sure we should be normalizing the casing here since URLs are case sensitive. As long as the request path doesn't match the redirect `Location` header, then we don't get into an infinite loop. I think it should be enough for us to compare the two right before calling `NextResponse.redirect`.

Reply:
That makes sense!